fuse/utils/interface_validator.py

```diff
@@ -0,0 +1,156 @@
+Unlike programming languages like C++, python does not have a built in way to enforce adherance to API when inheriting from a base class
+By using the code found in this script you can require the inheriting class to implement certain method and also control which arguments of the method signature is required.
+an actively used example can be seen in InsilicoLabModelBase() in insilico-lab-validation repo.
+For illustration, you can create a base class FruitBase (you can see related unit test in `fuse-med-ml/fuse/utils/tests/test_interface_validator.py`)
+
+
+```python
+from fuse.utils.interface_validator import InterfaceValidator, validate_signature
+
+class FruitBase(InterfaceValidator):
+    def __init__(self, *, wow='wow'):
+        super().__init__()
+        self.wow = wow
+
+    @validate_signature(args=['color'])  #only argument `color` will be validated to both exist and have the same type hint in the inheriting class
+    def draw(self, *, color:str = 'red', style:int=12, **kwargs):
+        print(f'FruitBase.draw: {color=} {style=}')
+
+    @validate_signature(args=['size', 'material']) #both args `size` and `material` will be validated
+    def print_3d(self, *, size:float = 2.7, material:str = 'iron'):
+        print(f'FruitBase.print_3d {size=} {material=}')
+```
+
+now, when you use a class that inherits from it, like:
+
+```python
+class Banana(FruitBase):
+    #...
+```
+
+the API will be enforced and exceptions will be raised if requirements are not met.
+
+
+#TODO:
+    1. allow validating return type hint
+    2. allow using @validate_signature without using any args, and that would mean that the exact same full signature is required
+    3. allow control over whether type hint is validated or not (is this needed?)
+"""
+
+
+def validate_signature(*, args: List[str] = None) -> Callable:
+    def wrapper(function: Callable) -> Any:
+        return validate_signature_impl(function, args=args)
+
+    return wrapper
+
+
+class validate_signature_impl(object):
+    def __init__(self, fn: Callable, args: List[str] = None) -> None:
+        self.fn = fn
+        self._args_to_validate = args
+
+    def __set_name__(self, owner: any, name: str) -> None:
+        # do something with owner, i.e.
+        curr_func = self.fn
+        print(f"decorating {curr_func} and using {owner}")
+        if not hasattr(owner, "_validate_methods"):
+            owner._validate_methods = {}
+        owner._validate_methods[name] = {}
+
+        signature_info = inspect.getfullargspec(self.fn)
+
+        for curr_arg in self._args_to_validate:
+            owner._validate_methods[name][curr_arg] = signature_info.annotations[
+                curr_arg
+            ]
+
+        # then replace ourself with the original method
+        setattr(owner, name, curr_func)
+
+
+class InterfaceValidator:
+    def __init__(self, verbose: bool = False):
+        self.__dict__["_properly_initialized_InterfaceValidator"] = True
```

Comment:
super().__init__() is recommded.
Why self.__dict__["_properly_initialized_InterfaceValidator"] and not self._properly_initialized_InterfaceValidator ?

Reply:
if you use `self._properly_initialized_InterfaceValidator` then you will immediately get into `self.__setattr__`  and get an exception that it doesn't exist yet

Reply:
"super().init() is recommded" - do you mean that I should switch from using `super().__init__()` to `super().init()` ?

Reply:
about super().__init__() ok I think I understand what you meant, that it's just generally a good practice, maybe even when you don't inherit from anything